src/Parallel/PhaseControl/PhaseControlTags.hpp

```diff
@@ -84,4 +91,100 @@ using reduction_data = Parallel::ReductionData<Parallel::ReductionDatum<
     tuples::tagged_tuple_from_typelist<TagsAndCombinesPresent>,
     TaggedTupleCombine>>;
 
+namespace OptionTags {
+/// Option tag for the collection of triggers that indicate synchronization
+/// points at which phase changes should be considered, and the associated
+/// `PhaseChange` objects for making the phase change decisions.
+///
+/// When the phase control is arbitrated on the main chare, the `PhaseChange`
+/// objects will be queried for their phase request in order of appearance in
+/// the nested list (i.e. first all of the `PhaseChange`s associated with the
+/// first trigger, in order, then those associated with the second trigger,
+/// etc.). The order therefore determines the order of resolution of
+/// simultaneous requests.
+///
+/// \note The nested collection types for this option tag gives the yaml
+/// format the slightly unusual form:
+///
+/// ```
+/// PhaseChangeAndTriggers:
+///   - - Trigger1
+///     - - PhaseChange1
+///       - PhaseChange2
+///   - - Trigger2
+///     - - PhaseChange3
+///       - PhaseChange4
+/// ```
+template <typename PhaseChangeRegistrars, typename TriggerRegistrars>
+struct PhaseChangeAndTriggers {
```

Comment:
Based on our discussion above I think it would be better to either not add this option tag in this PR, or add it in a very limited scope, e.g. in `tests/` or in a sample executable that uses it. The reason is that I think in our production executables we don't want to specify load-balancing, checkpointing etc. all in one big `PhaseChangeAndTriggers` list-of-pairs-of-triggers-and-lists. Note that I mean only this option tag here which describes the input-file interface; I think it's a reasonable design to actually store the phase changes in a data structure like this internally (defined by the DataBox tag below).

Reply:
I'm not sure I agree with this. Currently I'd rather have all the phase change info in one location in the input file. I don't think we have enough experience (basically zero right now) to merit a refactor of the design in this PR. I think it is much more important to get this feature and all of LB+checkpointing in than to spend time right now trying to decide on some optimal placement of the options in the input file when we have no idea yet what we will need and how the input file will look in a realistic case.

Reply:
Somewhat related to this, I'm realizing that I don't really understand what this structure means.  I think the goal here is to choose a phase to change to, but then why would it be useful to have more than one request on the same trigger?  Won't the first one (or the first one with a "do it now" request) always win?

Reply:
It would always 'win', but the information for the ones that are later on is still stored in the struct that determines the next phase to jump to. For instance, say we wanted to only have a global sync every 100 slabs, but every global sync we want to both checkpoint and load balance, in that order. The we would do:
```
PhaseChangeAndTriggers:
  - - Slabs:
        EvenlySpaced:
          Interval: 100
          Offset: 0
    - VisitAndReturn(Checkpoint)
    - VisitAndReturn(LoadBalance)
```

This is a reasonable set of instructions, and both things will happen as we would hope. Here's the summary of the control flow:
- trigger happens, data for both `PhaseChange`s are sent to Main
- elements halt, quiescence detected, control passes to Main
- Main 'arbitrates' the phase change, iterates to `VisitAndReturn(Checkpoint)`, finds the 'do it now' request, proceeds to checkpoint phase, mutating the main decision data as prescribed for `VisitAndReturn(Checkpoint)` (i.e. storing the 'return' info)
- quiescence after checkpoint, control passes to Main again
- Main 'arbitrates' the phase change again, first determining that it should return to `Evolve` phase but also continue considering other options, then getting the 'do it now' request from `LoadBalancing`, jumps to load balancing
- quiescence after load balancing, control passes to Main
- gets the return request back from the `VisitAndReturn(LoadBalance)` data, jumps back to `Evolve`

The same happens if they are in separate triggers that happen to go off at the same time, and they are executed in the reverse order if the order in the vector is reversed.
Does that clarify the ordering logic? I do confess that this is a bit subtle, and I'm happy to add to the documentation anything that would help make the reasoning here clearer.

Reply:
I understand now, thanks.  I think the documentation is fine and I just didn't read it carefully enough.

This ordering structure seems like a good argument for keeping all the phase changes together in the input file.  If they are separated it will not be clear what order the phases will execute in, and it will presumably not be possible to alter that order from the input file.

Reply:
I'm happy with merging this since it's important to get the functionality moving forward. That said, I would like to keep this input-file interface an open issue - I think (1) it's very hard to read and (2) it comes very close to "programming in the input-file". Just to reiterate point (1) I would like to point out that I think your snippet above misses a dash ;)

Reply:
I've opened issue #3030 for discussion of the input file interface for these features.

Reply:
thanks 👍